component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/component/AbstractMigrationHandler.java

```diff
@@ -0,0 +1,159 @@
+import java.util.function.Predicate;
+
+import org.talend.sdk.component.api.component.MigrationHandler;
+import org.talend.sdk.component.runtime.manager.spi.MigrationHandlerListenerExtension;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public abstract class AbstractMigrationHandler implements MigrationHandler {
+
+    private final Collection<MigrationHandlerListenerExtension> listeners = new CopyOnWriteArrayList<>();
+
+    @Getter
+    private Map<String, String> configuration;
+
+    /**
+     * @param incomingVersion the version of associatedData values.
+     * @param incomingData the data sent from the caller. Keys are using the path of the property as in component
+     * metadata.
+     *
+     * @return the set of properties for the current version.
+     */
+    @Override
+    public final Map<String, String> migrate(final int incomingVersion, final Map<String, String> incomingData) {
+        configuration = new HashMap<>(incomingData);
+        migrate(incomingVersion);
+
+        return Collections.unmodifiableMap(configuration);
+    }
+
+    /**
+     * @param incomingVersion the version of associated data values.
+     */
+    public abstract void migrate(final int incomingVersion);
+
+    /**
+     * @param listener the observer that will receive changes
+     */
+    public final synchronized void registerListener(final MigrationHandlerListenerExtension listener) {
+        log.debug("[registerListener] registering {}.", listener.getClass().getName());
+        listeners.add(listener);
+    }
+
+    /**
+     * @param listener the observer that will receive changes
+     */
+    public final synchronized void unRegisterListener(final MigrationHandlerListenerExtension listener) {
+        log.debug("[unRegisterListener] unregistering {}.", listener.getClass().getName());
+        listeners.remove(listener);
+    }
+
+    /**
+     * @param key
+     * @param value
+     */
+    public final void addKey(final String key, final String value) {
+        configuration.put(key, value);
+
+        listeners.forEach(e -> e.onAddKey(configuration, key, value));
+    }
+
+    /**
+     * @param oldKey configuration key existing
+     * @param newKey configuration key to rename
+     */
+    public final void renameKey(final String oldKey, final String newKey) {
+        configuration.put(newKey, configuration.get(oldKey));
+        configuration.remove(oldKey);
+
+        listeners.forEach(e -> e.onRenameKey(configuration, oldKey, newKey));
+    }
+
+    /**
+     * @param key configuration key
+     */
+    public final void removeKey(final String key) {
+        configuration.remove(key);
+
+        listeners.forEach(e -> e.onRemoveKey(configuration, key));
+    }
+
+    /**
+     * @param key configuration key
+     * @param newValue new value to set in migration
+     */
+    public final void changeValue(final String key, final String newValue) {
+        final String oldValue = configuration.get(key);
```

Comment:
could we have a Function instead of String for newValue ? but we let a string for the listener. so we could have something lilke:
public final void changeValue(final String key, final Function<String, String> func) {
	String new value = func.apply(config.get(key))
	configuration.put(key, newValue);
	listeners.forEach(e -> e.onChangeValue(configuration, key, oldValue, newValue));
}

Reply:
Ok the method exists after...

Reply:
Please check if key exists...